freqtrade/exchange/binance_public_data.py

```diff
@@ -0,0 +1,254 @@
+        # uploaded and have several hours delay
+        last_available_date = dt_now() - datetime.timedelta(days=2)
+        end = min(end, last_available_date)
+        if start >= end:
+            return DataFrame()
+        df = await _fetch_ohlcv(asset_type, symbol, timeframe, start, end, stop_on_404)
+        logger.debug(
+            f"Downloaded data for {pair} from https://data.binance.vision with length {len(df)}."
+        )
+    except Exception as e:
+        logger.debug("An exception occurred", exc_info=e)
+        df = DataFrame()
+
+    if not df.empty:
+        # only return the data within the requested time range
+        return df.loc[(df["date"] >= start) & (df["date"] < end)]
+    else:
+        return df
+
+
+def concat(dfs) -> DataFrame:
+    if all(df is None for df in dfs):
+        return DataFrame()
+    else:
+        return pd.concat(dfs)
+
+
+async def _fetch_ohlcv(
+    asset_type: str,
+    symbol: str,
+    timeframe: str,
+    start: datetime.date,
+    end: datetime.date,
+    stop_on_404: bool,
+) -> DataFrame:
+    # daily dataframes, `None` indicates missing data in that day (when `stop_on_404` is False)
+    dfs: list[DataFrame | None] = []
+    # the current day being processing, starting at 1.
+    current_day = 0
+
+    connector = aiohttp.TCPConnector(limit=100)
+    async with aiohttp.ClientSession(connector=connector, trust_env=True) as session:
+        # the HTTP connections has been throttled by TCPConnector
+        for dates in chunks(list(date_range(start, end)), 1000):
+            results = await asyncio.gather(
+                *(get_daily_ohlcv(asset_type, symbol, timeframe, date, session) for date in dates)
+            )
+            for result in results:
+                current_day += 1
+                if isinstance(result, Http404):
+                    if stop_on_404:
+                        # A 404 error on the first day indicates missing data
+                        # on https://data.binance.vision, we provide the warning and the advice.
+                        # https://github.com/freqtrade/freqtrade/blob/acc53065e5fa7ab5197073276306dc9dc3adbfa3/tests/exchange_online/test_binance_compare_ohlcv.py#L7
+                        if current_day == 1:
+                            logger.warning(
+                                "Failed to use fast download, fall back to rest API download, this "
+                                "can take more time. If you're downloading BTC/USDT:USDT, "
+                                "ETH/USDT:USDT, BCH/USDT:USDT, please first download "
+                                "data before 2020 (using `--timerange yyyymmdd-20200101`), and "
+                                "then download the full data you need."
+                            )
+                        logger.debug("Abort downloading from data.binance.vision due to 404")
+                        return concat(dfs)
```

Comment:
This message should definitely be higher than debug level.
It's rather important for us to diagnose "why" something it's taking forever

Reply:
Resovled the unnecessary download by replace `asyncio.gather` with `asyncio.create_task` and cancel the remaining tasks when we return:
[feat: cancel uncompleted tasks before return](https://github.com/freqtrade/freqtrade/pull/10858/commits/6b18c4f24c16213c52ec098863b787298322072e)

Added some user friendly warning messages when falling back to rest API:
[chore: add user friendly warnings](https://github.com/freqtrade/freqtrade/pull/10858/commits/8baa0f73103d86057195fa1a4dac74b8b30bd6e4)

Reply:
Resovled the unnecessary download by replace `asyncio.gather` with `asyncio.create_task` and cancel the remaining tasks when we return:
[feat: cancel uncompleted tasks before return](https://github.com/freqtrade/freqtrade/pull/10858/commits/6b18c4f24c16213c52ec098863b787298322072e)

Added some user friendly warning messages when falling back to rest API:
[chore: add user friendly warnings](https://github.com/freqtrade/freqtrade/pull/10858/commits/8baa0f73103d86057195fa1a4dac74b8b30bd6e4)